Support multiple schemas

vhdirk opened the pull request:
This PR adds initial support for multiple schemas in a single gschema file.
It's pretty basic and there's probably a lot of caveats, for instance:  

 - reusing flags/enums will cause these to be generated twice as aux tokenstream. Having a different key in each schema will make different flag structs/enums, which won't be compatible.

 - choices using the same key will not work

Is this something you would be willing to pull in?

Edit: I have some ideas to fix this fix the possible issues regarding flags and enums. I propose to to that in an additional PR to keep changesets small.

vhdirk committed:
multiple schemas

vhdirk committed:
fix tests

SeaDve commented on src/lib.rs:
```diff
@@ -302,14 +302,33 @@ pub fn gen_settings(
     let mut schemas = schema_list.schemas;
-    if schemas.len() > 1 {
-        emit_warning!(file_attr_span, "this macro only supports a single schema");
+    if schemas.len() == 0 {
```
I think the if statements could also be simplified with something like:

```rust
let schema = match schemas.as_ref() {
    [] => {
        // handle empty
    }
    [schema] => {
        schema.clone()
    }
    schemas if id_attr.is_none() {
        // some code
    }
    schemas => {
        // more code
    }
};

SeaDve commented:
Thanks for the PR! I did a quick review, but I'll do a full review next week, after my finals.

vhdirk committed:
fix remarks and add default impl condition

SeaDve commented:
> reusing flags/enums will cause these to be generated twice as aux tokenstream. Having a different key in each schema will make different flag structs/enums, which won't be compatible.

I'd like to have this fixed first before merging the MR.

> choices using the same key will not work

Not sure what you men by that

SeaDve commented on tests/io.github.seadve.test.multi.gschema.xml:
```diff
@@ -0,0 +1,99 @@
+<?xml version="1.0" encoding="utf-8"?>
```
I think having one of each keys for each child schema would be enough

vhdirk committed:
do not generate globals twice

vhdirk commented:
I've updated my PR so that it now promotes flags and enums to be standalone, decoupling them from individual schemas.
Since I haven't found any other (simple) way to do this, I've added a 'globals' attribute to the macro, that indicates whether or not they should be generated

SeaDve commented:
Looking good! 👍

> Since I haven't found any other (simple) way to do this, I've added a 'globals' attribute to the macro, that indicates whether or not they should be generated

Is there a specifc reason why would the user prefer not to generate globals?

vhdirk commented:
> Is there a specifc reason why would the user prefer not to generate globals?

Well, yes. Take the enum [here](https://github.com/SeaDve/gsettings-macro/blob/main/tests/io.github.seadve.test.gschema.xml#L8) for instance.
When generating multiple schemas, there's no way to tell of those global structures have been defined before. proc_macro's do not have the ability to look at previously generated code, and while you can keep a thread_local with information regarding previous invocations, you can't know the context that code was generated in. So, you have to set the flag manually.

The alternative would be to just generate everything in the schema in one go, but I've been working on that and it's quite a bit of work, TBH.

SeaDve commented:
Ah, that makes sense indeed.

I wonder if it is worth it to push through something like:

```rust
gen_settings! {
    FILE = "...";

    #[gen_settings(id = "...")]
    pub struct Settings;

    #[gen_settings(id = "...")]
    pub struct AnotherSettings;
}
```

It is a rough scratch of the API, but I think something like that would be less confusing and less thinking on the user side whether to generate the globals.

SeaDve commented:
We can also copy `quote_spanned` syntax:

```rust
gen_settings! {"path/to/schema/file"=> 
    const FILE = "...";

    #[gen_settings(id = "...")]
    pub struct Settings;

    #[gen_settings(id = "...")]
    pub struct AnotherSettings;
}
```
